### PR TITLE
cranelift: Use requested ISA Flags in run tests

### LINKED ISSUE #1891: Test specific flags in Cranelift filetests
In #1880 @bnjbvr brought up an issue with how Cranelift's filetests use (or don't use) ISA-specific flags. He was trying to use `target x86_64 use_new_backend` in a `test run` filetest but the `SingleFunctionCompiler` doesn't know anything about ISA-specific `Flags`, only shared `Flags`. (Perhaps these two sets of flags should be merged in some way, but what I will propose next doesn't require that).

The current behavior is:
 - check if the default host architecture (used by `SingleFunctionCompiler`) is the same as the one specified by the `target` directive; if not, skip the test (though in the future this should print some sort of `IGNORED [filename]` so the user knows what happened, see #1558)
 - compile and run the test using the default host `TargetIsa`

I propose we change the behavior to:
 - check if the default host `TargetIsa` _is compatible with_ the one specified by the `target` directive; if not, skip the test, etc.
 - compile and run the test using the `target` directive's `TargetIsa`--this will use any special flags assigned by the user, e.g. `use_new_backend`

To get this "_is compatible with_" behavior, I propose we add a new `Flags::matches` function to the shared `Flags` and to each ISA-specific `Flags` (this likely has to be done as generated code in `gen_settings.rs` so it only needs to be done once). Then, we expose this function as `TargetIsa::matches` and implement it in each ISA as `self.flags.matches(&other.flags) && self.isa_flags.matches(&other.isa_flags)`. This way we can compare TargetIsa's for compatibility.

How does `Flags::matches` work? I think it needs to iterate over the field descriptors and compare them both to the default value and to the other value. Remember that `matches` only goes one way: if `a` matches `b` it does not necessarily mean that `b` matches `a`. Though it passes some simple tests, I'm not 100% confident that the following is correct so I would appreciate feedback:

```
fn get_bit(byte: u8, bit: u8) -> bool {
    let mask = 1 << bit;
    (byte & mask) != 0
}

impl Flags {
    fn matches(&self, other: &Self) -> bool {
        let shared_default = settings::Flags::new(settings::builder());
        let default = Flags::new(&shared_default, builder());

        // Check each detail until we see presets.
        let mut byte_offset = 0;
        for d in &DESCRIPTORS {
            byte_offset = d.offset as usize;
            match d.detail {
                detail::Detail::Bool { bit } => {
                    let self_bit = get_bit(self.bytes[byte_offset], bit);
                    let default_bit = get_bit(default.bytes[byte_offset], bit);
                    let other_bit = get_bit(other.bytes[byte_offset], bit);
                    if self_bit != default_bit && self_bit != other_bit {
                        return false;
                    }
                }
                detail::Detail::Num | detail::Detail::Enum { .. } => {
                    let self_byte = self.bytes[byte_offset];
                    let default_byte = default.bytes[byte_offset];
                    let other_byte = other.bytes[byte_offset];
                    if self_byte != default_byte && self_byte != other_byte {
                        return false;
                    }
                }
                detail::Detail::Preset => break,
            }
        }

        // Then check each preset bit.
        for byte_offset in byte_offset + 1..self.bytes.len() {
            for bit in 0..8 {
                let self_bit = get_bit(self.bytes[byte_offset], bit);
                let default_bit = get_bit(default.bytes[byte_offset], bit);
                let other_bit = get_bit(other.bytes[byte_offset], bit);
                if self_bit != default_bit && self_bit != other_bit {
                    return false;
                }
            }
        }

        true
    }
}
```

### PR DESCRIPTION
**Pull request by afonso360**
👋 Hey,

This PR Enables using the ISA flags that run tests request. We skip tests that request flags that the host arch does not support.

I did a quick pass on the backends to find ops that we have optimizations based on flags and enabled the tests for those.

Fixes #1891

### CODE REVIEW

File: cranelift/filetests/src/test_run.rs [hunk 1]
```diff
@@ -24,6 +25,73 @@ pub fn subtest(parsed: &TestCommand) -> anyhow::Result<Box<dyn SubTest>> {
...
+            .and_then(|val| val.as_bool())
+            .unwrap_or(false);
+
+        if requested && !available_in_host {
+            return Err(format!(
+                "skipped {}: host does not support ISA flag {}",
+                context.file_path, req_value.name
+            ));
+        }
+    }
```
Comment (jameysharp) [inline cranelift/filetests/src/test_run.rs:90]: It seems surprising to me to silently ignore non-bool flags here. How do you feel about this?

```suggestion
    for req_value in requested_flags {
        if let Some(requested) = req_value.as_bool() {
            let available_in_host = host
                .isa_flags()
                .iter()
                .find(|val| val.name == req_value.name)
                .and_then(|val| val.as_bool())
                .unwrap_or(false);

            if requested && !available_in_host {
                return Err(format!(
                    "skipped {}: host does not support ISA flag {}",
                    context.file_path, req_value.name
                )); 
            }
        } else {
            unimplemented!("ISA flag {} of kind {:?}", req_value.name, req_value.kind());
        }
    }
```

**Review by jameysharp — COMMENTED** [review]
Tell me if I've understood the implications of this change correctly:

It looks like the existing behavior is to select an ISA with all native-supported features enabled.

The new behavior is to select an ISA with all features _disabled_ unless you specifically ask for them, and then skip the test if the features you asked for aren't available.

Is that right?

That seems like a good move to me: I'm not sure I like the idea that the same tests check different things depending on which host machine you run them on. Being explicit that some tests are useful to run both with and without particular hardware features seems great.

But I'd like to get @alexcrichton or somebody to double-check, or @cfallin when he's back next week.

File: cranelift/filetests/src/test_run.rs [hunk 2]
```diff
@@ -24,6 +25,73 @@ pub fn subtest(parsed: &TestCommand) -> anyhow::Result<Box<dyn SubTest>> {
...
+) -> Box<dyn TargetIsa> {
+    let mut builder = cranelift_native::builder_with_options(infer_native_flags)
+        .expect("Unable to build a TargetIsa for the current host");
+
+    for value in isa_flags {
+        match value.kind() {
+            SettingKind::Bool => builder
+                .set(value.name, &format!("{}", value.as_bool().unwrap()))
+                .unwrap(),
+            _ => unimplemented!(),
```
[the remark this answers is not in the document]
Comment (afonso360) [reply cranelift/filetests/src/test_run.rs:44]: I'm going to backport what I did in #4453 which adds a `value_string` function and makes this a whole lot cleaner, and supports all flag kinds.

**Comment by afonso360** [comment]
>It looks like the existing behavior is to select an ISA with all native-supported features enabled.
>
> The new behavior is to select an ISA with all features disabled unless you specifically ask for them, and then skip the test if the features you asked for aren't available.
>
> Is that right?

Yes, except that we don't start with *all* features disabled, we start with the defaults enabled by cranelift. So on x86_64 for example we start with `has_sse41` and a bunch of others. These can be explicitly disabled to test the lowering without those features.

**Review by alexcrichton — COMMENTED** [review]
Seems reasonable to me!

We technically already had a problem where if Cranelift was buggy missing a feature flag then if you ran the tests on that host without the feature flag tests would fail where tests run on a host with the feature flag would not fail. Overall this seems like a better state of things where tests are run independent of the host executing the test, and if a test fails then it should fail on any host which can execute the test.

**Review by jameysharp — APPROVED** [review]
Great, I'm going to merge this then. Separately we can experiment with going further by turning off all feature flags by default during these tests, but I'm guessing that will involve more noise, updating a bunch of tests to enable features that were default before.

**Pull request merged**